ucp/_libs/ucp_tiny.pyx

```diff
@@ -0,0 +1,329 @@
+        ucp_params.features     = UCP_FEATURE_TAG | UCP_FEATURE_WAKEUP | UCP_FEATURE_STREAM
+        ucp_params.request_size = sizeof(ucp_request)
+        ucp_params.request_init = ucp_request_init
+        status = ucp_config_read(NULL, NULL, &config)
+        assert_error(status == UCS_OK)
+        
+        status = ucp_init(&ucp_params, config, &self.context)
+        assert_error(status == UCS_OK)
+        
+        worker_params.field_mask  = UCP_WORKER_PARAM_FIELD_THREAD_MODE
+        worker_params.thread_mode = UCS_THREAD_MODE_MULTI
+        status = ucp_worker_create(self.context, &worker_params, &self.worker)
+        assert_error(status == UCS_OK)
+
+        cdef int ucp_epoll_fd
+        status = ucp_worker_get_efd(self.worker, &ucp_epoll_fd)
+        assert_error(status == UCS_OK)
+
+        self.epoll_fd = epoll_create(1)
+        cdef epoll_event ev
+        ev.data.fd = ucp_epoll_fd
+        ev.events = EPOLLIN 
+        cdef int err = epoll_ctl(self.epoll_fd, EPOLL_CTL_ADD, ucp_epoll_fd, &ev)
+        assert_error(err == 0)
+
+        ucp_config_release(config)
+        printf("ApplicationContext() - self.worker: %p\n", self.worker)
+
+    
+    def create_listener(self, callback_func, port = None):
+        self._bind_epoll_fd_to_event_loop()
+        if port is None:
+            # Ref https://unix.stackexchange.com/a/132524
+            import socket
+            s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            s.bind(('', 0))
+            port = s.getsockname()[1]
+            s.close()
+        
+        cdef _listener_callback_args *args = <_listener_callback_args*> malloc(sizeof(_listener_callback_args))
+        args.ucp_worker = self.worker
+        args.py_func = <void*> callback_func
+        Py_INCREF(callback_func)
+
+        cdef ucp_listener_params_t params = c_util_get_ucp_listener_params(port, _listener_callback, <void*> args)
+        print("create_listener() - Start listening on port %d" % port)
+        listener = Listener(port)
+        cdef ucs_status_t status = ucp_listener_create(self.worker, &params, &listener._ucp_listener)
+        c_util_get_ucp_listener_params_free(&params)
+        assert_error(status == UCS_OK)
+        return listener
+
+
+    async def create_endpoint(self, str ip_address, port):
+        self._bind_epoll_fd_to_event_loop()
+        ret = Endpoint()
+        ret._ucp_worker = self.worker
+        cdef ucp_ep_params_t params = c_util_get_ucp_ep_params(ip_address.encode(), port)
+        cdef ucs_status_t status = ucp_ep_create(self.worker, &params, &ret._ucp_ep)
+        c_util_get_ucp_ep_params_free(&params)
+        assert_error(status == UCS_OK)
+                    
+        import numpy as np
+        ret.unique_send_tag = np.uint64(hash(uuid.uuid4()))
+        ret.unique_recv_tag = np.uint64(hash(uuid.uuid4()))
+        tags = np.array([ret.unique_recv_tag, ret.unique_send_tag], dtype="uint64")
+        await ret.stream_send(tags, tags.nbytes)
```

Comment:
Is `stream_send` something we can `await` on? Also do we need to handle a returned result from it?

Reply:
I use `stream_send()` for the initial exchange of tag IDs

Reply:
And in UCX, it is non-blocking :)